File: src/main/java/com/numbers/discordbot/commands/JoinVoiceCommand.java
package com.numbers.discordbot.commands;

import com.numbers.discordbot.*;
import java.util.*;
import sx.blah.discord.handle.impl.events.guild.channel.message.*;
import sx.blah.discord.handle.obj.*;
import sx.blah.discord.util.*;
import com.numbers.discordbot.filter.MessageFilter;

@Command
public class JoinVoiceCommand {

    @Command
    @MessageFilter(mentionsBot = true, 
            regex = ".*join.*",
            eventType = MentionEvent.class)
    public void handle(MentionEvent event)  
    {
        MessageTokenizer tokenizer = event.getMessage().tokenize();
        tokenizer.nextMention(); //bot
        tokenizer.nextWord(); //join
        
        handleCommand(event, tokenizer);
    }
    
    @Command
    @MessageFilter(eventType = 
            MessageEvent.class, 
            prefixCheck = true, 
            startsWith = "join")
    public void handle(MessageEvent event){
        MessageTokenizer tokenizer = event.getMessage().tokenize();
        tokenizer.nextWord(); //command
        tokenizer.nextWord(); //join
        
        handleCommand(event, tokenizer);
    }
    
    private void handleCommand(MessageEvent event, MessageTokenizer tokenizer){
        if(tokenizer.hasNext()){
            if(tokenizer.hasNextMention()){
                MessageTokenizer.MentionToken token = tokenizer.nextMention();
                String content = token.getContent();
                content = content.substring(3, content.length() - 2);
                IUser mention = event.getGuild().getUserByID(Long.parseLong(content));
                if(mention != null){
                    joinUser(event, mention);
                }else{
                    joinUser(event, event.getAuthor());
                }
            }else{
                String channelName = tokenizer.nextWord().getContent();
                System.out.println(channelName);
                List<IVoiceChannel> channels = event.getGuild().getVoiceChannelsByName(channelName);
                if(!channels.isEmpty()){
                    joinChannel(event, channels.get(0));
                }else{
                    joinUser(event, event.getAuthor());
                }
            }
        }else{
            joinUser(event, event.getAuthor());
        }
    }

    private void joinUser(MessageEvent event, IUser user)
    {
        IVoiceChannel channel = event.getGuild().getVoiceChannels().stream()
                .filter(chnl -> chnl.getConnectedUsers().contains(user))
                .findAny().orElse(event.getGuild().getVoiceChannels().get(0));

        joinChannel(event, channel);
    }

    private void joinChannel(MessageEvent event, IVoiceChannel channel)
    {
        if (event.getGuild().getConnectedVoiceChannel() != channel) {
            if(event.getGuild().getConnectedVoiceChannel() != null)
                event.getGuild().getConnectedVoiceChannel().leave();
            channel.join();
        }
    }

}

File: src/main/java/com/numbers/discordbot/commands/JoinVoiceCommand.kt
package com.numbers.discordbot.commands

import com.numbers.discordbot.Command
import com.numbers.discordbot.filter.MessageFilter
import sx.blah.discord.handle.impl.events.guild.channel.message.MentionEvent
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageEvent
import sx.blah.discord.handle.obj.IGuild
import sx.blah.discord.handle.obj.IUser
import sx.blah.discord.handle.obj.IVoiceChannel
import sx.blah.discord.util.MessageTokenizer

@Command(name = "Join Voice")
class JoinVoiceCommand {

    @Command
    @MessageFilter(mentionsBot = true, regex = ".*join.*", eventType = MentionEvent::class, readableUsage = "join {@<User>||\$channelName}")
    fun handleMention(event: MentionEvent){
        val tokenizer = event.message.tokenize()
        tokenizer.nextMention()
        tokenizer.nextWord()

        join(event, tokenizer)
    }

    @Command
    @MessageFilter(startsWith = "join", prefixCheck = true, eventType = MessageEvent::class, readableUsage = "join {@<User>||\$channelName}")
    fun handlePrefix(event : MessageEvent){
        val tokenizer = event.message.tokenize()
        tokenizer.nextWord()
        tokenizer.nextWord()

        join(event, tokenizer)
    }

    private fun join(event : MessageEvent, tokenizer: MessageTokenizer){
        when {
            tokenizer.hasNextMention() -> joinMention(event, tokenContent = tokenizer.nextWord().content)
            tokenizer.hasNextWord() -> joinChannelName(event, tokenContent = tokenizer.nextWord().content)
            else -> join(event.author, event.guild)
        }
    }

    private fun joinChannelName(event: MessageEvent, tokenContent: String){
        event.guild.getVoiceChannelsByName(tokenContent).first().let { it.join() }
    }

    private fun joinMention(event: MessageEvent, tokenContent: String){
        val mentionId = tokenContent.substring( 3, tokenContent.length - 1).toLong()

        print("id: ${mentionId}")

        event.guild.getUserByID(mentionId).let { join(it, event.guild) }
    }

    private fun join(user: IUser, guild: IGuild){
        val channel : IVoiceChannel =
                guild.voiceChannels.find { it.connectedUsers.contains(user) }
                ?: guild.voiceChannels.first()

        channel.let { join(it, guild) }
    }

    private fun join(channel: IVoiceChannel, guild: IGuild){
        if(guild.connectedVoiceChannel != null && guild.connectedVoiceChannel != channel){
            guild.connectedVoiceChannel.leave()
        }
        channel.join()
    }
}